cdap-tms/src/main/java/co/cask/cdap/messaging/store/hbase/HBaseMessageTable.java

```diff
@@ -0,0 +1,234 @@
+        putBuilder.add(columnFamily, TX_COL, txPtrBytes);
+      } else {
+        putBuilder.add(columnFamily, PAYLOAD_COL, entry.getPayload());
+        if (entry.isTransactional()) {
+          putBuilder.add(columnFamily, TX_COL, Bytes.toBytes(entry.getTransactionWritePointer()));
+        }
+      }
+      batchPuts.add(putBuilder.build());
+    }
+    if (!batchPuts.isEmpty()) {
+      getHTable().put(batchPuts);
+    }
+  }
+
+  @Override
+  public void delete(TopicId topicId, long transactionWritePointer) throws IOException {
+    byte[] rowPrefix = MessagingUtils.toBytes(topicId);
+    byte[] stopRow = Bytes.stopKeyForPrefix(rowPrefix);
+    Scan scan = tableUtil.buildScan()
+      .setStartRow(rowPrefix)
+      .setStopRow(stopRow)
+      .build();
+
+    try (ResultScanner scanner = getHTable().getScanner(scan)) {
+      Iterator<Result> iterator = scanner.iterator();
+      Result result = iterator.next();
+      if (result.containsColumn(columnFamily, TX_COL)) {
+        byte[] txCol = result.getValue(columnFamily, TX_COL);
+        if (Bytes.equals(txCol, Bytes.toBytes(transactionWritePointer))) {
+          Delete delete = tableUtil.buildDelete(result.getRow())
+            .build();
+          getHTable().delete(delete);
+        }
+      }
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (hTable != null) {
+      hTable.close();
+    }
+  }
+
+  private HTable getHTable() throws IOException {
+    if (hTable != null) {
+      return hTable;
+    }
+
+    synchronized (this) {
+      if (hTable != null) {
+        return hTable;
+      }
+      hTable = tableUtil.createHTable(hConf, tableId);
+      return hTable;
+    }
+  }
+
+  private static class HBaseCloseableIterator extends AbstractCloseableIterator<Entry> {
+    private final ResultScanner scanner;
+    private final Transaction transaction;
+    private final byte[] colFamily;
+    private int limit;
+    private boolean closed = false;
+
+    HBaseCloseableIterator(ResultScanner scanner, int limit, byte[] colFamily, @Nullable Transaction transaction) {
+      this.scanner = scanner;
+      this.transaction = transaction;
+      this.colFamily = colFamily;
+      this.limit = limit;
+    }
+
+    @Override
+    protected Entry computeNext() {
+      if (closed) {
+        return endOfData();
+      }
+
+      if (limit <= 0) {
+        close();
+        return null;
+      }
+
+      while (true) {
+        Result result = null;
+        try {
+          result = scanner.next();
+        } catch (IOException e) {
+          Throwables.propagate(e);
+        }
+
+        if (result == null) {
+          break;
+        }
+        MessageTable.Entry entry = new MessageEntry(result.getRow(), result.getValue(colFamily, PAYLOAD_COL),
+                                                    result.getValue(colFamily, TX_COL));
+        if (validEntry(entry, transaction)) {
+          limit--;
+          return entry;
+        }
```

Comment:
If it is not valid entry, you should signal the end of this iterator (by returning `endOfData`, otherwise it will consume pass the uncommitted message, messing up the ordering.